### PR TITLE
xcpretty: format additional build steps.

### PR DESCRIPTION
**Pull request by alon-k**
Add formatting for a few more build steps.
@byohay

### CODE REVIEW

**Comment by alon-k** [comment]
Fixed cucumber issue. Forgot to run integration tests and had a small inconsistency between formatter.rb and simple.rb APIs.

File: spec/xcpretty/formatters/simple_spec.rb [hunk 1]
```diff
@@ -67,6 +67,12 @@ module XCPretty
         "[foo] > Compiling asset catalog sample/Assets.xcassets"
       end
 
+      it "formats sym link output" do
+        @formatter.format_sym_link('Source.h',
```
Comment (byohay) [inline spec/xcpretty/formatters/simple_spec.rb]: Add folder path to source.

File: spec/fixtures/constants.rb [hunk 2]
```diff
@@ -626,6 +626,12 @@
     builtin-copy -exclude .DS_Store -exclude CVS -exclude .svn -exclude .git -exclude .hg -resolve-src-symlinks /Users/jenkins/.jenkins/workspace/temp/Repo/third_party/ReactiveObjC/ReactiveObjC/UISegmentedControl+RACSignalSupport.h /Users/jenkins/.jenkins/workspace/temp/Repo/test_package/Debug-iphonesimulator/ReactiveObjC.framework/Headers
 )
 
+SAMPLE_SYM_LINK = %Q(
+SymLink /Users/jenkins/Library/Developer/Xcode/DerivedData/Blah-fsncgmwwwwqfhufpjvhbxnyuqkxp/Build/Intermediates.noindex/ArchiveIntermediates/Blah/BuildProductsPath/Release-iphoneos/SomeFramework.framework /Users/jenkins/Library/Developer/Xcode/DerivedData/Blah-fsncgmwwwwqfhufpjvhbxnyuqkxp/Build/Intermediates.noindex/ArchiveIntermediates/Blah/IntermediateBuildFilesPath/UninstalledProducts/iphoneos/SomeFramework.framework (in target: SomeFramework)
```
Comment (byohay) [inline spec/fixtures/constants.rb]: Better to use a different name for the target so that the test will be more precise.

File: lib/xcpretty/parser.rb [hunk 3]
```diff
@@ -167,12 +167,9 @@ module Matchers
...
 
     # @regex Captured groups
-    # $1 = file
-    PROCESS_PCH_MATCHER = /^ProcessPCH\s.*\s(.*.pch)/.freeze
-
-    # @regex Captured groups
-    # $1 file_path
-    PROCESS_PCH_COMMAND_MATCHER = /^\s*.*\/usr\/bin\/clang\s.*\s\-c\s(.*)\s\-o\s.*/.freeze
+    # $1 = nil or ++ (C or C++)
+    # $2 = file
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: `file_path`

File: lib/xcpretty/parser.rb [hunk 4]
```diff
@@ -167,12 +167,9 @@ module Matchers
...
+    PROCESS_PCH_MATCHER = /^ProcessPCH(\+\+)?\s.*\s(.*.pch)/.freeze
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: For literal dot use `\.`.

File: lib/xcpretty/formatters/formatter.rb [hunk 5]
```diff
@@ -38,7 +38,7 @@ def format_passing_test(suite, test, time);                EMPTY; end
     def format_pending_test(suite, test);                      EMPTY; end
     def format_measuring_test(suite, test, time);              EMPTY; end
     def format_failing_test(suite, test, reason, file_path);   EMPTY; end
-    def format_process_pch(file, build_target);                EMPTY; end
+    def format_process_pch(cpp, file, build_target);           EMPTY; end
```
Comment (byohay) [inline lib/xcpretty/formatters/formatter.rb]: I prefer that the formatter will get the language and not just the postfix, that is, the parser will add the `C`. Then the interface between the parser and the formatter are nicer.

File: lib/xcpretty/formatters/formatter.rb [hunk 6]
```diff
@@ -38,7 +38,7 @@ def format_passing_test(suite, test, time);                EMPTY; end
...
     def format_process_pch_command(file_path, build_target);   EMPTY; end
```
Comment (byohay) [inline lib/xcpretty/formatters/formatter.rb:42]: It exists in the JSON formatter, this is where it is used. The question is whether we completely disregard the JSON formatter for now, or we should try to break it as less as possible. @barakwei?

File: lib/xcpretty/parser.rb [hunk 7]
```diff
@@ -64,6 +64,23 @@ module Matchers
...
+    COMPILE_SWIFT_SOURCES_MATCHER = /^CompileSwiftSources\s.*$/.freeze
+
+    # @regex Captured groups
+    COMPILE_SWIFT_MATCHER = /^CompileSwift\s.*$/.freeze
+
+    # @regex Captured groups
+    SWIFT_CODE_GENERATION_MATCHER = /^SwiftCodeGeneration\s.*$/.freeze
+
+    # @regex Captured groups
+    # $1 = bytecode file
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: `bytecode file path`

File: lib/xcpretty/parser.rb [hunk 8]
```diff
@@ -64,6 +64,23 @@ module Matchers
...
+    SWIFT_CODE_GENERATION_COMMAND_MATCHER = /^\s*.*\/usr\/bin\/swift.*\s-c\s.*\s(.*\.bc)/.freeze
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: 1. @barakwei this is not the primary command, do we really need this?
2. Regex can be dangerously slow! For example the sample code takes 80k steps to match: https://regex101.com/r/6Anrlz/1. We already had performance issues with xcpretty, so we want to keep it fast.
There are several ways to do it: In this case, using `.*` or anything similar is expensive, you should use avoid it when you can. For example: You don't need `^\s*.*` at the beginning, and I'm not sure that you need to find `-c` also, in which case you can merge the two `.*`. Please check your other regexes as well.

File: lib/xcpretty/formatters/simple.rb [hunk 9]
```diff
@@ -44,6 +44,27 @@ def format_compile(file_name, file_path, build_target)
...
+    def format_swift_code_generation(build_target)
+      format("Generating object files from bytecode", build_target: build_target)
+    end
+
+    def format_swift_code_generation_command(file_path)
+      format("Generating object file from", File.basename(file_path))
+    end
+
+    def format_copy_swift_libs(app_path, build_target)
+      format("Copying required Swift libs for", File.basename(app_path),
```
Comment (byohay) [inline lib/xcpretty/formatters/simple.rb]: Why required? Aren't all copied libs required?

File: lib/xcpretty/parser.rb [hunk 10]
```diff
@@ -46,6 +46,24 @@ module Matchers
     # $3 configuration
     CLEAN_TARGET_MATCHER = /^=== CLEAN TARGET\s(.*)\sOF PROJECT\s(.*)\sWITH CONFIGURATION\s(.*)\s===/.freeze
 
+    # @regex Captured groups
+    # $1 = app file
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: Specify weather this is just the name or path: `app file path/name`. Same for other places.

File: lib/xcpretty/parser.rb [hunk 11]
```diff
@@ -303,7 +303,7 @@ module Warnings
 
       # @regex Captured groups
       # $1 = whole warning
-      GENERIC_WARNING_MATCHER = /^warning:\s(.*)$/.freeze
+      GENERIC_WARNING_MATCHER = /^.*[Ww]arning:\s(.*)$/.freeze
```
Comment (byohay) [inline lib/xcpretty/parser.rb:332]: You can simplify by doing: `[Ww]arning:\s(.*)$`

**Review by byohay — CHANGES_REQUESTED** [review]
Round.

**Comment by byohay** [comment]
@alon-k also fix PR title.

File: lib/xcpretty/parser.rb (hunk 8 above)
Comment (barakwei) [reply lib/xcpretty/parser.rb]: I don't think we need it, but @alon-k insisted, so I leave the decision to you.

File: spec/xcpretty/formatters/simple_spec.rb (hunk 1 above)
Comment (alon-k) [reply spec/xcpretty/formatters/simple_spec.rb]: Done

File: spec/fixtures/constants.rb (hunk 2 above)
Comment (alon-k) [reply spec/fixtures/constants.rb]: Done

File: lib/xcpretty/formatters/simple.rb (hunk 9 above)
Comment (alon-k) [reply lib/xcpretty/formatters/simple.rb]: They are. This says that the libs which are copied (and thus **required**) are being copied right now. The wording can be changed if you'd like.

File: lib/xcpretty/parser.rb (hunk 3 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: Fixed

File: lib/xcpretty/parser.rb (hunk 4 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: Fixed

File: lib/xcpretty/formatters/formatter.rb (hunk 5 above)
Comment (alon-k) [reply lib/xcpretty/formatters/formatter.rb]: Changed

File: lib/xcpretty/formatters/formatter.rb (hunk 6 above)
Comment (alon-k) [reply lib/xcpretty/formatters/formatter.rb:42]: I'll put it back for now, but would also like to know what @barakwei thinks about this

File: lib/xcpretty/parser.rb (hunk 7 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: Fixed

File: lib/xcpretty/parser.rb (hunk 8 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: 1. I think that either this should be included, or xcpretty shouldn't print anything at all for SwiftCodeGeneration logs, as otherwise they contain no meaningful information other than the target and the number of object files being generated (notice that the file path is only included in the command, and not in the SwiftCodeGeneration log - btw, @barakwei didn't you want to open an issue about that?)
2. I had to find -c because otherwise it matches other /usr/bin/swift commands from earlier in the log (in the other swift build commands) which are unwanted as they compile a bunch of files at the same time and are just hard to work with in general.
Regarding the `^\s*.*` in the beginning, I've removed it. This cut down most of the steps (only takes 5k steps to match now), so I think the performance hit of the -c isn't very significant.

File: lib/xcpretty/parser.rb (hunk 11 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb:332]: Notice that this will make the libpng test case fail as it will cause the regex to capture "Warning: blah blah" instead of "blah blah" in that case. It does make the regex faster though. We can let it slide for the sake of performance, or leave it as it is - your call.

File: lib/xcpretty/parser.rb (hunk 10 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: Done

File: lib/xcpretty/parser.rb (hunk 8 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: edit: I checked and the -c barely costs any steps anyways. Removing it brings it down from 5k to 4.6k steps at the cost of matching things we don't want to match

**Comment by alon-k** [comment]
Round complete.
PR Title fixed.

File: lib/xcpretty/parser.rb (hunk 8 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: improved further (to around 1k steps) in latest push.

Comment (alon-k) [reply lib/xcpretty/parser.rb]: improved to around 400 steps in latest push

File: lib/xcpretty/formatters/simple.rb (hunk 9 above)
Comment (alon-k) [reply lib/xcpretty/formatters/simple.rb]: Fixed.

File: lib/xcpretty/formatters/formatter.rb (hunk 6 above)
Comment (alon-k) [reply lib/xcpretty/formatters/formatter.rb:42]: Also, if we are talking performance: the regex that matches this is one of the worst ones in terms of performance (my 80k step regex before we fixed it was based on it), which means that even though we don't use it, since we left it in it causes a hit to performance. It might be a good idea to remove it after all. Might also be a good idea to go over the other regexes and optimize them, and also remove everything we don't use, since the parser still attempts to match even the regexes that we don't use for simple formatter and removing them might also improve performance.

File: lib/xcpretty/formatters/simple.rb (hunk 9 above)
Comment (alon-k) [reply lib/xcpretty/formatters/simple.rb]: Changed in latest push.

File: lib/xcpretty/parser.rb (hunk 11 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb:332]: Left as is for now. The performance hit isn't very significant here either - 100 steps instead of 30 steps.

**Comment by alon-k** [comment]
Pushed some more fixes.

File: lib/xcpretty/formatters/formatter.rb (hunk 6 above)
Comment (alon-k) [reply lib/xcpretty/formatters/formatter.rb:42]: Spoke to Barak and removed it.

File: lib/xcpretty/formatters/formatter.rb (hunk 5 above)
Comment (byohay) [inline lib/xcpretty/formatters/formatter.rb]: Change to `language`.

File: lib/xcpretty/parser.rb [hunk 12]
```diff
@@ -64,6 +82,23 @@ module Matchers
...
     COMPILE_COMMAND_MATCHER = /^\s*(.*clang\s.*\s\-c\s(.*\.(?:m|mm|c|cc|cpp|cxx))\s.*\.o)$/.freeze
 
+    # @regex Captured groups
+    COMPILE_SWIFT_SOURCES_MATCHER = /^CompileSwiftSources\s.*$/.freeze
+
+    # @regex Captured groups
+    COMPILE_SWIFT_MATCHER = /^CompileSwift\s.*$/.freeze
+
+    # @regex Captured groups
+    SWIFT_CODE_GENERATION_MATCHER = /^SwiftCodeGeneration\s.*$/.freeze
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: Consider using state of build_target, then you can get rid of this regex altogether.

**Review by byohay — CHANGES_REQUESTED** [review]
F&S.

File: lib/xcpretty/formatters/formatter.rb (hunk 5 above)
Comment (alon-k) [reply lib/xcpretty/formatters/formatter.rb]: Done

File: lib/xcpretty/parser.rb (hunk 12 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: Done

**Comment by alon-k** [comment]
Pushing last fix + additional swift related formatting that was discovered while testing other projects. Please wait before reviewing

**Comment by alon-k** [comment]
Done. Please review.

File: spec/xcpretty/parser_spec.rb [hunk 13]
```diff
@@ -456,13 +456,22 @@ module XCPretty
...
-        "Foo.entitlements",
-        "MobileApp.app.xcent",
+        "\"Foo.entitlements\"",
+        "MobileApp.build/MobileApp.app.xcent",
         "foo"
       )
       @parser.parse(SAMPLE_PROCESS_PRODUCT_PACKAGING)
     end
 
+    it "parses process product packaging #2" do
```
Comment (byohay) [inline spec/xcpretty/parser_spec.rb]: `parses process product packaging without entitlements`

File: lib/xcpretty/parser.rb [hunk 14]
```diff
@@ -275,9 +275,9 @@ module Matchers
     MKDIR_MATCHER = /^MkDir\s(?:.*\/)?(.+)$/.freeze
 
     # @regex Captured groups
-    # $1 = entitlements
+    # $1 = entitlements / provisioning + directory they are in
     # $2 = package name
-    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s"((?:\\"|[^"])*)"\s.+\/(.+)$/.freeze
+    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s((?:\\\s|\S)*)\s.+\/(.+\/.+)$/.freeze
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: Put `"` as optional.

Comment (byohay) [inline lib/xcpretty/parser.rb]: Why did you change `.+` to `.+\/.+`?

File: lib/xcpretty/parser.rb [hunk 15]
```diff
@@ -275,9 +275,9 @@ module Matchers
     MKDIR_MATCHER = /^MkDir\s(?:.*\/)?(.+)$/.freeze
 
     # @regex Captured groups
-    # $1 = entitlements
+    # $1 = entitlements / provisioning + directory they are in
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: Unclear what you mean.

**Review by byohay — CHANGES_REQUESTED** [review]
F&S.

File: lib/xcpretty/parser.rb (hunk 15 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: Fixed

File: lib/xcpretty/parser.rb (hunk 14 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: Changed back

Comment (alon-k) [reply lib/xcpretty/parser.rb]: Done

File: spec/xcpretty/parser_spec.rb (hunk 13 above)
Comment (alon-k) [reply spec/xcpretty/parser_spec.rb]: Fixed

**Comment by alon-k** [comment]
Done.

File: lib/xcpretty/parser.rb [hunk 16]
```diff
@@ -271,9 +271,9 @@ module Matchers
     MKDIR_MATCHER = /^MkDir\s(?:.*\/)?(.+)$/.freeze
 
     # @regex Captured groups
-    # $1 = entitlements
+    # $1 = entitlements / provisioning profile name
```
Comment (byohay) [inline lib/xcpretty/parser.rb]: name -> path

**Review by byohay — CHANGES_REQUESTED** [review]
F&S.

File: lib/xcpretty/parser.rb (hunk 16 above)
Comment (alon-k) [reply lib/xcpretty/parser.rb]: Fixed

**Comment by alon-k** [comment]
Done.

**Review by byohay — APPROVED** [review]
